### DISCUSSION TITLE
2.4.1 to 2.6.1 migration fails with "ModuleNotFoundError: No module named 'fido2.features'"

### DISCUSSION THREAD

**Discussion by pmatuszy** [healthchecks/healthchecks#786] [General]
Hi
I've decided to upgrade from 2.4.1 to 2.6.1.
I did the manage.py migrate etc
when I start gunicorn in the log file I get this:

"ModuleNotFoundError: No module named 'fido2.features'"

fido2 is installed in v-env
I decided to reinstall it with
![image](https://user-images.githubusercontent.com/58336955/215841739-21f21a89-a1ed-4c00-a320-767aee35c7f8.png)

![image](https://user-images.githubusercontent.com/58336955/215841819-d2b94446-b433-4bfc-b3e5-bccddf0facf3.png)
but it didn't help
![image](https://user-images.githubusercontent.com/58336955/215841995-0557b6b1-2f21-4fbd-8bdc-9e72b1ef2182.png)

Have you got any idea how to fix it?
Any clues are welcome.
Paul

**Comment by cuu508** [comment]
To me it looks like fido2==1.1.0 got installed in one virtualenv, but manage.py is using another virtualenv.

With the correct virtualenv activated the following should work:

```
$ python
Python 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fido2.features
>>> 
```

**Comment by pmatuszy** [comment]
Hi
I've check it .

In my script just before running gunicorn, I've added 
pip3 list -v
python   (to run python interactively with import you suggested)
![image](https://user-images.githubusercontent.com/58336955/216049723-b6a1fe6f-d239-44ae-b5b9-4b8deba98558.png)

so it uses the same virtual environment for sure... but gunicorn get this error seen with tail -F .... command on errorlog

![image](https://user-images.githubusercontent.com/58336955/216050014-099c8285-7772-4571-b76a-c561145ce8c5.png)

I am not sure how to approach the problem - the env looks good but when I run gunicorn it doesn't see fido2 ...

**Comment by cuu508** [comment]
Try this and check if you see the expected version and filesystem path:

```
$ python
Python 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fido2
>>> fido2.__version__
'1.1.0'
>>> fido2.__file__
'/home/user/venvs/healthchecks/lib/python3.10/site-packages/fido2/__init__.py'
>>> 
```

**Comment by pmatuszy** [comment]
ok  - here you are
![image](https://user-images.githubusercontent.com/58336955/216051550-f1e6eee6-c50e-4b22-8805-f45471c3221e.png)

**Comment by cuu508** [comment]
What are the contents of run-gunicorn-server.sh?

Is gunicorn also installed inside the virtualenv, and are you using the gunicorn that is installed in the virtualenv?

If you do the following with no virtualenv activated, what happens?

```
$ python
Python 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import fido2
>>> fido2.__version__
'1.1.0'
>>> fido2.__file__
'/home/user/venvs/healthchecks/lib/python3.10/site-packages/fido2/__init__.py'
>>> 
```

**Comment by pmatuszy** [comment]
ok - here is it - I've removed the values but you can see the env and how I set it up

```
$ cat ./run-gunicorn-server.sh
export PING_EMAIL_DOMAIN=
export EMAIL_HOST=
export EMAIL_HOST_PASSWORD=
export EMAIL_HOST_USER=
export EMAIL_PORT=587
export EMAIL_USE_TLS=True
export EMAIL_USE_VERIFICATION=False     # https://healthchecks.io/docs/self_hosted_configuration/#EMAIL_USE_VERIFICATION
export PING_BODY_LIMIT=90000000
export REGISTRATION_OPEN=False
export SITE_NAME=
export DEFAULT_FROM_EMAIL=
export PING_EMAIL_DOMAIN=
export PING_ENDPOINT=
export REGISTRATION_OPEN=False
export SITE_NAME=


export SITE_ROOT=''

export SIGNAL_CLI_SOCKET=
export SIGNAL_CLI_SOCKET=

export TWILIO_ACCOUNT=
export TWILIO_AUTH=
export TWILIO_FROM=
export TWILIO_USE_WHATSAPP=True

export ALLOWED_HOSTS='*'
export DB=postgres

# https://docs.djangoproject.com/en/3.1/ref/settings/#conn-max-age
export CONN_MAX_AGE=200
export DB_HOST=
export DB_NAME=
export DB_USER=
export DB_PASSWORD=
export DB_PORT=
# require               # https://www.postgresql.org/docs/10/libpq-connect.html#LIBPQ-CONNECT-SSLMODE
export DB_SSLMODE=
export SECRET_KEY=

export DEBUG=False

export RP_ID=

cd /encrypted/webapps/healthchecks

python3 -m venv hc-venv
source hc-venv/bin/activate


pip3 list -v

# https://docs.gunicorn.org/en/stable/settings.html

GUNICORN_CMD_ARGS="--config /encrypted/healthchecks/gunicorn.conf.py"

opt_config="--config /encrypted/healthchecks/gunicorn.conf.py"
opt_error_log="--error-logfile /tmp/gunicorn.errorlog"
opt_access_log="--access-logfile /tmp/gunicorn.accesslog"

log_level=error
log_level=info
log_level=debug

python

/encrypted/webapps/hc-venv/bin/python /encrypted/webapps/hc-venv/bin/gunicorn $opt_config hc.wsgi:application $opt_error_log $opt_access_log

```

**Comment by cuu508** [comment]
There might be a mixup between two venvs:

```
cd /encrypted/webapps/healthchecks

python3 -m venv hc-venv
source hc-venv/bin/activate
pip3 list -v
```

This lists the packages in venv located at `/encrypted/webapps/healthchecks/hc-venv`

```
/encrypted/webapps/hc-venv/bin/python /encrypted/webapps/hc-venv/bin/gunicorn ....
```

This runs from a venv located at `/encrypted/webapps/hc-venv/`.

**Comment by pmatuszy** [comment]
yes! indeed. I messed up 2 venvs. I've corrected it and set to "/encrypted/webapps/healthchecks/hc-venv" and now it all works!

I am flooded now with email and signal messages :-)
Thank you for your help!
Paul
